matrix_sdk/src/models/message.rs

```diff
@@ -0,0 +1,274 @@
+    msgs: Vec<MessageWrapper>,
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub struct MessageWrapper(MessageEvent);
+
+impl Deref for MessageWrapper {
+    type Target = MessageEvent;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl PartialEq for MessageWrapper {
+    fn eq(&self, other: &MessageWrapper) -> bool {
+        self.0.event_id == other.0.event_id
+            && self.0.room_id == other.0.room_id
+            && self.0.origin_server_ts == other.0.origin_server_ts
+            && self.0.sender == other.0.sender
+            && self.0.content == other.0.content
+    }
+}
+
+impl Eq for MessageWrapper {}
+
+impl PartialOrd for MessageWrapper {
+    fn partial_cmp(&self, other: &MessageWrapper) -> Option<Ordering> {
+        Some(self.0.origin_server_ts.cmp(&other.0.origin_server_ts))
+    }
+}
+
+impl Ord for MessageWrapper {
+    fn cmp(&self, other: &MessageWrapper) -> Ordering {
+        self.partial_cmp(other).unwrap_or(Ordering::Equal)
+    }
+}
+
+impl PartialEq for MessageQueue {
+    fn eq(&self, other: &MessageQueue) -> bool {
+        self.msgs.len() == other.msgs.len()
+            && self
+                .msgs
+                .iter()
+                .zip(other.msgs.iter())
+                .all(|(msg_a, msg_b)| msg_a.event_id == msg_b.event_id)
+    }
+}
+
+impl MessageQueue {
+    /// Create a new empty `MessageQueue`.
+    pub fn new() -> Self {
+        Self {
+            msgs: Vec::with_capacity(20),
+        }
+    }
+
+    /// Inserts a `MessageEvent` into `MessageQueue`, sorted by by `origin_server_ts`.
+    ///
+    /// Removes the oldest element in the queue if there are more than 10 elements.
+    pub fn push(&mut self, msg: MessageEvent) -> bool {
+        // only push new messages into the queue
+        if let Some(latest) = self.msgs.last() {
+            if msg.origin_server_ts < latest.origin_server_ts {
+                return false;
+            }
+        }
+
+        let message = MessageWrapper(msg);
+        match self.msgs.binary_search_by(|m| m.cmp(&message)) {
+            Ok(pos) => self.msgs.insert(pos, message),
+            Err(pos) => self.msgs.insert(pos, message),
+        }
+        if self.msgs.len() > 10 {
+            self.msgs.remove(0);
+        }
+        true
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &MessageWrapper> {
+        self.msgs.iter()
+    }
+}
+
+impl IntoIterator for MessageQueue {
+    type Item = MessageWrapper;
+    type IntoIter = IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.msgs.into_iter()
+    }
+}
+
+pub(crate) mod ser_deser {
+    use super::*;
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<MessageQueue, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let messages: Vec<EventJson<MessageEvent>> = de::Deserialize::deserialize(deserializer)?;
+
+        // TODO this should probably bail out if deserialization fails not skip the message
```

Comment:
Probably bailing out is the right idea. A de-serialization error here means that the cache is corrupted so people might wan to rebuild it somehow.